js.go

```diff
@@ -846,21 +839,195 @@ func (sub *Subscription) ConsumerInfo() (*ConsumerInfo, error) {
 	return js.getConsumerInfo(stream, consumer)
 }
 
-func (sub *Subscription) Poll() error {
+type pullOpts struct {
+	noWait  bool
+	ttl     time.Duration
+	ctx     context.Context
+	expires time.Duration
+}
+
+type PullOpt interface {
+	configurePull(opts *pullOpts) error
+}
+
+// pullOptFn is a function option used to configure pull consumers.
+type pullOptFn func(opts *pullOpts) error
+
+func (opt pullOptFn) configurePull(opts *pullOpts) error {
+	return opt(opts)
+}
+
+// PullNoWait enables receiving a no messages error from the server
+// in case there are no new messages when making a pull request.
+func PullNoWait() PullOpt {
+	return pullOptFn(func(opts *pullOpts) error {
+		opts.noWait = true
+		return nil
+	})
+}
+
+// PullExpiresIn defines when to let the server expire an inflight pull request.
+func PullExpiresIn(duration time.Duration) PullOpt {
+	return pullOptFn(func(opts *pullOpts) error {
+		opts.expires = duration
+		return nil
+	})
+}
+
+// PullMaxWaiting defines the max inflight pull requests to be delivered more messages.
+func PullMaxWaiting(n int) SubOpt {
+	return subOptFn(func(opts *subOpts) error {
+		opts.cfg.MaxWaiting = n
+		return nil
+	})
+}
+
+// Fetch pulls a batch of messages from a stream for a pull consumer.
+func (sub *Subscription) Fetch(batch int, opts ...PullOpt) ([]*Msg, error) {
+	if sub == nil {
+		return nil, ErrBadSubscription
+	}
+
+	var o pullOpts
+	for _, opt := range opts {
+		if err := opt.configurePull(&o); err != nil {
+			return nil, err
+		}
+	}
+
+	// Default is no wait, so async.
+	if o.ctx != nil && o.ttl != 0 {
+		return nil, ErrContextAndTimeout
+	}
+
 	sub.mu.Lock()
-	if sub.jsi == nil || sub.jsi.deliver != _EMPTY_ || sub.jsi.pull == 0 {
+	if sub.jsi == nil || !sub.jsi.pull {
 		sub.mu.Unlock()
-		return ErrTypeSubscription
+		return nil, ErrTypeSubscription
 	}
-	batch := sub.jsi.pull
-	nc, reply := sub.conn, sub.Subject
+
+	// FIXME: We don't need the original inbox from the subscriber anymore,
+	// maybe each consumer should have its own mux for responses?
+	nc, _ := sub.conn, sub.Subject
 	stream, consumer := sub.jsi.stream, sub.jsi.consumer
 	js := sub.jsi.js
+
+	ttl := o.ttl
+	if ttl == 0 {
+		ttl = js.wait
+	}
 	sub.mu.Unlock()
 
-	req, _ := json.Marshal(&nextRequest{Batch: batch})
+	// PullNoWait when enabled returns:
+	// - 404 right away if there are no more messages,
+	// - 409 in case it reached MaxAckPending.
+	//
+	// PullExpiresIn will give a duration to the server to expire our pull request
+	// to this inbox only and will serve the newest one.
+	//
+	// Default is to let the pull requests linger until there are 512 waiting requests,
```

Comment:
so we are trying to do some sort of server protection for over pulling, which will eventually fail when 512 pull requests have been pending. This makes no sense as it is not related in any way to the pull arguments but to some unexpected consumer configuration that is not related to the number of live consumers. When the 512 hits, the subscription should be expected to fail? This is bizarre and complex. From a client perspective, it either gets messages and waits, or the request times out. It would be nice if the server responded to the requestor with a 404 after the expiration triggers (this still requires the client to keep some sort of timer).

To me, it seems that the issue of multiple consumers in a queue group should be handled by using a workqueue stream. In this case, the consumer can pull what it wants, and once handled it goes away. Each consumer is independent.

IMHO this is overly complex, and the pull subscriber should simply poll, get at most the messages requested, and repeated. If the behavior of the polling is that it shouldn't be too frequent it should be documented. The double configuration is really unexpected and overcomplicates the functionality.